Работа по 15 спринту

UdachiTomo opened the pull request.

UdachiTomo committed:
Работа по 15 спринту

mmbabaev commented on CoreData/MainCoreData.swift:
```diff
@@ -0,0 +1,42 @@
+    case Error
+}
+
+final class MainCoreData {
```
хорошо, что в отдельном классе

mmbabaev commented on CoreData/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,202 @@
+    
+    convenience override init() {
+        let context = MainCoreData.shared.context
+        try! self.init(context: context)
```
здесь было бы хорошо обработать ошибку

mmbabaev commented on CoreData/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,202 @@
+        delegate?.store(
+            self,
+            didUpdate: TrackerCategoryStoreUpdate(
+                insertedIndexes: insertedIndexes!,
```
интересно, что используете IndexSet-ы, но так делать force unwrap нехорошо. Тут легко его хотя бы на пустой сет заменить изначально.

В больших проектах бывает так: сейчас они всегда заполнены, завтра придет новый человек, добавит еще какую-то логику между вызовами методов под условиями, где присывается nil для этих свойств.

mmbabaev commented on CoreData/MainCoreData.swift:
```diff
@@ -0,0 +1,42 @@
+import Foundation
+import CoreData
+
+enum MainCoreDataEror: Error {
```
опечатка

mmbabaev commented on CoreData/MainCoreData.swift:
```diff
@@ -0,0 +1,42 @@
+import CoreData
+
+enum MainCoreDataEror: Error {
+    case Error
```
case называют с маленькой буквы.  Также то, что это Error понятно по названию класса

mmbabaev commented on CoreData/TrackerStore.swift:
```diff
@@ -0,0 +1,56 @@
+                       color: color.color,
+                       emoji: emoji,
+                       schedule: schedule.compactMap { WeekDay(rawValue: $0)}
+                       )
```
скачет форматирование.
Предлагаю использовать `cmd + A`, `ctrd + I` , так будет легче читать код

Для информации: в реальных проектах часто используют линтеры/форматеры, типа библиотек SwiftLint / SwiftFormatter

mmbabaev commented on Models/TrackerCategory.swift:
```diff
@@ -3,4 +3,12 @@ import Foundation
     let title: String
     let trackers: [Tracker]
+    
+//    func visibleTrackers(filterString: String) -> [Tracker] {
```
не оставляй закомменченный код

mmbabaev commented on TrackerApp/TrackerViewController/TrackersViewController.swift:
```diff
@@ -185,12 +186,38 @@ final class TrackersViewController: UIViewController, UITextFieldDelegate, Choos
+    func createTracker(_ tracker: Tracker, categoryTitle: String) {
+        var categoryToUpdate: TrackerCategory?
+        let categories: [TrackerCategory] = trackerCategoryStore.trackerCategories
+        for i in 0..<categories.count {
```
`categoyToUpdate = categories.first { $0.title == categoryTitle }`

mmbabaev commented on TrackerApp/TrackerViewController/TrackersViewController.swift:
```diff
@@ -185,12 +186,38 @@ final class TrackersViewController: UIViewController, UITextFieldDelegate, Choos
+                categoryToUpdate = categories[i]
+            }
+        }
+        if categoryToUpdate != nil {
```
if let

mmbabaev commented on TrackerApp/TrackerViewController/TrackersViewController.swift:
```diff
@@ -185,12 +186,38 @@ final class TrackersViewController: UIViewController, UITextFieldDelegate, Choos
-        updateVisibleCategories(visibleCategories)
+        setDayOfWeek()
+        updateCategories()
+        completedTrackers = try! self.trackerRecordStore.fetchTrackerRecord()
```
снова форс анрап

UdachiTomo committed:
2.1 Исправлены замечания ревьювера
2.2 Реализована фукнция блокировки кнопки, если не выбраны все параметры при создании трекера

mmbabaev commented on CoreData/TrackerCategoryStore.swift:
```diff
@@ -163,10 +168,10 @@ extension TrackerCategoryStore: NSFetchedResultsControllerDelegate {
-                deletedIndexes: deletedIndexes!,
-                updatedIndexes: updatedIndexes!,
-                movedIndexes: movedIndexes!
+                insertedIndexes: [],
```
вообще странно тут сделано, я имел в виду, что insertedIndexes. и соседние массивы можно инициализировать пустым массивом при создании свойства. Тогда опционал будет вообще не нужен. 

В текущей реализации все IndexSet не нужны

UdachiTomo committed:
2.3 Произведена работа над force unwrap

UdachiTomo committed:
3. Исправлены небольшие ошибки

mmbabaev commented on TrackerApp/CreateNewTrackerViewController/CreateNewTrackerTableViewCell.swift:
```diff
@@ -19,39 +19,30 @@ final class CreateNewTrackerTableViewCell: UITableViewCell {
+                secondLabel.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 16)
             ])
-        } else if ((secondLabel.text?.isEmpty) != nil) {
+        } else if secondLabel.text?.isEmpty != false {
```
тут вроде снова не совсем верно, если text nil, то в условие не попадает. 
Можно так привести и уже проверять isEmpty: `(secondLabel.text ?? "").isEmpty`

mmbabaev commented on TrackerApp/CreateNewTrackerViewController/CreateNewTrackerViewController.swift:
```diff
@@ -340,24 +340,24 @@ extension CreateNewTrackerViewController: UICollectionViewDataSource {
             if selectedEmojiCell != nil {
-                collectionView.deselectItem(at: selectedEmojiCell!, animated: true)
-                collectionView.cellForItem(at: selectedEmojiCell!)?.backgroundColor = .white
+                collectionView.deselectItem(at: selectedEmojiCell ?? [0], animated: true)
```
здесь можно пользоваться конструкцией if let
